File: examples/rrm/rrm.cpp
#include <TMB.hpp>


template<class Type>
bool isNA(Type x){
  return R_IsNA(asDouble(x));
}

template<class Type>
bool isFinite(Type x){
  return R_finite(asDouble(x));
}






template<class Type>
Type objective_function<Type>::operator() ()
{
	// DATA SECTION
	DATA_VECTOR(x);		 		// log annual escapement.
	DATA_VECTOR(tot_harvest); 	// total harvest (sum over sectors)
	DATA_ARRAY(i_count);       // weir or aerial count index.
	// DATA_MATRIX(aerial); 		// peak aerial count (year, tributary)
	DATA_ARRAY(wcpue);  		// weekly test fishery cpue (year,week)
	DATA_ARRAY(in_river);		// in river mark-recapture data.

	// PARAMETER SECTION
	PARAMETER(mu);
	PARAMETER(sigma);
	PARAMETER_VECTOR(log_escapement);
	PARAMETER_VECTOR(log_q);
	PARAMETER_VECTOR(log_sig_i);
	

	int nyrs = tot_harvest.size();
	int nobs = i_count.dim(1);
	vector<Type> Et = exp(log_escapement);
	vector<Type> q  = exp(log_q);
	vector<Type> Ct(nyrs);
	matrix<Type> ihat(nyrs,nobs);

	Type objfun = 0;
	Type nll_count = 0;
	Type nll_river = 0;
	for (int i = 0; i < nyrs; ++i)
	{
		
		// negative loglikelihood for count obs.
		for (int j = 0; j < nobs; ++j)
		{
			if( !isNA(i_count(i,j)) )
			{
				Type obs = log(i_count(i,j));
				Type pre = log(q(j) * Et(i));
				Type sig = exp(log_sig_i(j));
				nll_count -= dnorm(obs,pre,sig,true);
				
			}
		}

		// negative loglikeilhood for Inriver MR estimates
		if( !isNA(in_river(i,0)) )
		{
			Type obs = log(in_river(i,0));
			Type pre = log(Et(i)+tot_harvest(i));
			Type sig = sqrt(1.0+in_river(i,1)/(in_river(i,0)*in_river(i,0)));
			nll_river -= dnorm(obs,pre,sig,true);
		}
	}

	std::cout<<nll_river<<"\n";

	Type f;
	f = -sum(dnorm(x,mu,sigma,true));
	// f = -sum(dnbinom(x,mu,sigma,TRUE));

	ADREPORT(sigma);
	REPORT(sigma);

	objfun = nll_count + nll_river;
	return objfun;

}
//



template<class Type>
Type arrival_model()
{
	
}

File: examples/rrm/rrm.R



# --------------------------------------------------------------------------- #
# --------------------------------------------------------------------------- #
# Grab Kuskokwim Chinook data and tailor it to my needs.
data_file  <- 'Kusko_estimates2015_Dec22.csv'
kusko.data <- read.csv(data_file,header=T, na.string='NA')

# Modifications
Year <- as.vector(kusko.data$Year)
nyr  <- length(Year)

# Harvest
harvest <- as.matrix(kusko.data[substr(names(kusko.data),1,2)=="H."])
tot.harvest <- rowSums(harvest,na.rm=TRUE)

# Weir counts
weir <- as.matrix(kusko.data[substr(names(kusko.data),1,2)=="w."])

# Aerial counts
aerial <- as.matrix(kusko.data[substr(names(kusko.data),1,2)=="a."])

# In.river
in.river <- as.matrix(kusko.data[substr(names(kusko.data),1,3)=="In."])


# Number of scaling coefficients.
nq <- dim(weir)[2] + dim(aerial)[2]
i_count <- cbind(weir,aerial)

# Test fishery weekly cpue
wcpue <- as.matrix(kusko.data[substr(names(kusko.data),1,3)=="rpw"])

# --------------------------------------------------------------------------- #
# --------------------------------------------------------------------------- #





# --------------------------------------------------------------------------- #
# --------------------------------------------------------------------------- #
# RUN RECONSTRUCTION MODEL.  
# AUTHORS: Steven Martell, Hamachan, Haflinger
# 	- Use a run-timing model to predict arrivals (A_t) and departures (D_t)
# 	- N_t = E_y \left[ \int_{t=0}^{t} f(\mu_d,\sigma_m)dt 
#           -\int_{t=0}^{t-s}g(\mu_d,\sigma_m)dt \right]
# 	- Bethel test fishery starts on June 1, each year and runs through Aug 24.
#   - June 1 corresponds to week 22.
# 
# 	- For the beta distribution:
# 		b = [(E(x)-1)(E(x)^2-E(x)+V(x))]V(x)
# 		a = (E(x)b)/(1-E(x))
# 	TODO LIST:
#  [ ] Build Simulation Model.
#  [ ] Run simulation testing.
# 
# --------------------------------------------------------------------------- #
# --------------------------------------------------------------------------- #
# library(lubridate)
# Mean Escapement
meanRunSize <- 100
days        <- seq(1,365,by=1) / 365

# 
# Mean Arrival date (day of year) 
mu.aday <- lubridate::yday(as.Date("1978-06-16")) / 365

# Stdev in Arrival date (days)
sd.aday <- 35 / 365

# Survey life (s in days)
s <- 42 /365

# Arrival model parameters
mu  = mu.aday
var = sd.aday*sd.aday
b = ( (mu-1)*(mu^2-mu+var) )/var
a = ( mu * b )/(1-mu)

# Cumulative arrival model
avec <- pbeta(days,a,b)

# Cumulative death model
dvec <- pbeta(days-s,a,b)

# Daily run size
nt   <- meanRunSize * (avec - dvec)

# sample dates for geting mean cpue
idate <- seq(lubridate::yday("1970-06-01"),by=7,length=10)

# test fishery cpue
set.seed(3424)
sig.epsilon <- 0.12
epsilon     <- rnorm(length(idate),0,sig.epsilon)
it          <- nt[idate]*exp(epsilon)












# rrm.R
# R-script for run reconstruction model (rrm)

n = 100
x = rnorm(n,mean=0,sd=1)


library(TMB)
compile("rrm.cpp",flags=" -O0")
dyn.load(dynlib("rrm"))

print(weir)
# weir[is.na(weir)]<-0
inputData <- list(x=x,tot_harvest=tot.harvest,
                  i_count=i_count,wcpue=wcpue,
                  in_river=in.river)

params <- list(mu=0,sigma=1,log_escapement=rep(0,length=nyr),
               log_q=rep(0,length=nq),log_sig_i=rep(0,length=nq))

# Construct an R object that represents the C++ function
# f = MakeADFun(inputData,parameters=params)#,random=c("mu"))
f = MakeADFun(inputData,parameters=params,DLL="rrm")#,random=c("mu"))

# The objective function value
f$fn()

# The gradient (first order derivatives)
f$gr()

# The Hessian (second order derivatives)
f$he()

# Likelihood maximization
fit <- nlminb(f$par,f$fn,f$gr,lower=c(-10.0,0.0),upper=c(10.0,10.0))



